File: app/src/main/java/com/digicoffer/lauditor/Appointments/Adapters/AppointmentHistoryAdapter.java
package com.digicoffer.lauditor.Appointments.Adapters;

import android.content.Context;
import android.text.Editable;
import android.text.TextWatcher;
import android.view.LayoutInflater;
import android.view.View;
import android.view.ViewGroup;
import android.widget.EditText;
import android.widget.ImageView;
import android.widget.LinearLayout;
import android.widget.TextView;

import androidx.annotation.NonNull;
import androidx.appcompat.widget.AppCompatButton;
import androidx.recyclerview.widget.RecyclerView;

import com.digicoffer.lauditor.Appointments.Models.AppointmentModel;
import com.digicoffer.lauditor.R;
import com.digicoffer.lauditor.CommonFiles.GlobalFiles.AndroidUtils;

import org.json.JSONObject;

import java.text.ParseException;
import java.text.SimpleDateFormat;
import java.util.ArrayList;
import java.util.Date;
import java.util.Locale;

public class AppointmentHistoryAdapter extends RecyclerView.Adapter<AppointmentHistoryAdapter.ViewHolder> {

    private ArrayList<AppointmentModel> appointmentList;
    private Context context;
    private OnNoteActionListener noteActionListener;
    String noteId = "";

    public interface OnNoteActionListener {
        void onSaveNote(AppointmentModel appointment, String note, int position);

        void onCancelNote(AppointmentModel appointment, int position);

        void onEditNote(AppointmentModel appointment, String noteId, String noteText, int position);

        void onDeleteNote(AppointmentModel appointment, String noteId, int position);
    }

    public AppointmentHistoryAdapter(ArrayList<AppointmentModel> appointmentList, Context context, OnNoteActionListener listener) {
        this.appointmentList = appointmentList;
        this.context = context;
        this.noteActionListener = listener;
    }

    @NonNull
    @Override
    public ViewHolder onCreateViewHolder(@NonNull ViewGroup parent, int viewType) {
        View view = LayoutInflater.from(parent.getContext()).inflate(R.layout.item_appointment_history, parent, false);
        return new ViewHolder(view);
    }

    @Override
    public void onBindViewHolder(@NonNull ViewHolder holder, int position) {
        AppointmentModel appointment = appointmentList.get(position);

        // Hide timeline line for last item
        if (position == appointmentList.size() - 1) {
            holder.timelineLine.setVisibility(View.INVISIBLE);
        } else {
            holder.timelineLine.setVisibility(View.VISIBLE);
        }

        // Set appointment title (count from end)
        int appointmentNumber = appointmentList.size() - position;
        String title = getOrdinal(appointmentNumber) + " Appointment";
        holder.tvAppointmentTitle.setText(title);

        // Format and set date/time
        String formattedDateTime = formatDateTime(appointment.getAppointment_from(), appointment.getAppointment_to());
        holder.tvAppointmentDateTime.setText(formattedDateTime);

        // Set payment info
        if (appointment.getPayment() != null) {
            String paymentInfo = "Payment: " + appointment.getPayment().getLabel();
            holder.tvPaymentInfo.setText(paymentInfo);
        }

        // Set status
        holder.tvStatusBadge.setText(appointment.getAppointment_status());
        // Set status badge AND timeline dot color
        setStatusBadgeStyle(holder.tvStatusBadge, holder.timelineDot, appointment.getAppointment_status());

        // Handle note section visibility
        if (appointment.isNoteExpanded()) {
            holder.llNoteSection.setVisibility(View.VISIBLE);
            holder.btnAddNote.setVisibility(View.GONE);
            holder.etNoteInput.setText(appointment.getTempNote());
            holder.etNoteInput.setSelection(holder.etNoteInput.getText().length());
        } else {
            holder.llNoteSection.setVisibility(View.GONE);
            holder.btnAddNote.setVisibility(View.VISIBLE);
        }

        // Add Note button click
        holder.btnAddNote.setOnClickListener(v -> {
            appointment.setNoteExpanded(true);
            notifyItemChanged(position);
        });

        holder.llNotesContainer.removeAllViews();

        if (appointment.isNoteExpanded()) {
            // Editing Mode → Hide previous notes
            holder.llNotesContainer.setVisibility(View.GONE);
            holder.btnAddNote.setVisibility(View.GONE);

        } else if (appointment.getNotes() != null && appointment.getNotes().length() > 0) {

            // Show previous notes only if NOT in edit mode
            holder.llNotesContainer.setVisibility(View.VISIBLE);
            holder.btnAddNote.setVisibility(View.GONE);


            for (int i = 0; i < appointment.getNotes().length(); i++) {

                try {
                    JSONObject noteObj = appointment.getNotes().getJSONObject(i);

                    String noteText = noteObj.optString("note");
                    String createdOn = noteObj.optString("created_on");

                    View noteView = LayoutInflater.from(context)
                            .inflate(R.layout.appointments_notes_item, holder.llNotesContainer, false);

                    TextView tvNoteText = noteView.findViewById(R.id.tv_note_text);
                    TextView tvNoteDate = noteView.findViewById(R.id.tv_note_date);

                    tvNoteText.setText(noteText);
                    tvNoteDate.setText(formatNoteDate(createdOn));
                    ImageView ivEdit = noteView.findViewById(R.id.iv_edit_note);
                    ImageView ivDelete = noteView.findViewById(R.id.iv_delete_note);

                    noteId = noteObj.optString("id");

                    String localNoteId = noteObj.optString("id");

                    ivEdit.setOnClickListener(v -> {
                        appointment.setNoteExpanded(true);
                        appointment.setEditingNote(true);
                        appointment.setEditingNoteId(localNoteId);
                        appointment.setTempNote(noteText);
                        notifyItemChanged(position);
                    });

                    ivDelete.setOnClickListener(v -> {

                        AndroidUtils.showConfirmationDialog(context, "Delete Note", "Are you sure you want to delete this note?", new AndroidUtils.OnConfirmListener() {
                                    @Override
                                    public void onSave() {
//                                    dialog.dismiss();
                                        if (noteActionListener != null) {
                                            noteActionListener.onDeleteNote(appointment, localNoteId, position);
                                        }
                                    }

                                    @Override
                                    public void onCancel() {
//                                dialog.dismiss();
//                                ViewMembersData();
                                    }
                                }
                        );
                    });

                    holder.llNotesContainer.addView(noteView);

                } catch (Exception e) {
                    e.printStackTrace();
                }
            }
        }
//        String note = holder.etNoteInput.getText().toString().trim();
//        if (note.isEmpty()) {
            AndroidUtils.ToggleButton(0, holder.btnSaveNote);
//        } else {
//            AndroidUtils.ToggleButton(1, holder.btnSaveNote);
//        }

        // Character counter
        holder.etNoteInput.addTextChangedListener(new TextWatcher() {
            @Override
            public void beforeTextChanged(CharSequence s, int start, int count, int after) {
            }

            @Override
            public void onTextChanged(CharSequence s, int start, int before, int count) {
                int length = s.length();
                holder.tvCharCounter.setText(length + "/500");
                appointment.setTempNote(s.toString());
                if (s.toString().isEmpty()) {
                    AndroidUtils.ToggleButton(0, holder.btnSaveNote);
                } else {
                    AndroidUtils.ToggleButton(1, holder.btnSaveNote);
                }
            }

            @Override
            public void afterTextChanged(Editable s) {
            }
        });

        // Cancel button
        holder.btnCancelNote.setOnClickListener(v -> {
            appointment.setNoteExpanded(false);
            appointment.setTempNote("");
            holder.etNoteInput.setText("");
            notifyItemChanged(position);

            if (noteActionListener != null) {
                noteActionListener.onCancelNote(appointment, position);
            }
        });

        holder.btnSaveNote.setOnClickListener(v -> {
            String noteText = holder.etNoteInput.getText().toString().trim();
            if (noteText.isEmpty()) return;

            if (appointment.isEditingNote()) {
                // EDIT existing note
                if (noteActionListener != null) {
                    noteActionListener.onEditNote(
                            appointment,
                            appointment.getEditingNoteId(),
                            noteText,
                            position
                    );
                }
            } else {
                // NEW note
                if (noteActionListener != null) {
                    noteActionListener.onSaveNote(
                            appointment,
                            noteText,
                            position
                    );
                }
            }

            // Reset UI
            appointment.setNoteExpanded(false);
            appointment.setEditingNote(false);
            appointment.setEditingNoteId("");
            appointment.setTempNote("");

            holder.etNoteInput.setText("");
            notifyItemChanged(position);
        });

    }

    @Override
    public int getItemCount() {
        return appointmentList.size();
    }

    private String formatDateTime(String fromDateTime, String toDateTime) {
        try {
            SimpleDateFormat inputFormat = new SimpleDateFormat("yyyy-MM-dd'T'HH:mm:ss", Locale.ENGLISH);
            SimpleDateFormat dateFormat = new SimpleDateFormat("MMM dd, yyyy", Locale.ENGLISH);
            SimpleDateFormat timeFormat = new SimpleDateFormat("hh:mm a", Locale.ENGLISH);

            Date fromDate = inputFormat.parse(fromDateTime);
            Date toDate = inputFormat.parse(toDateTime);

            if (fromDate != null && toDate != null) {
                String date = dateFormat.format(fromDate);
                String fromTime = timeFormat.format(fromDate);
                String toTime = timeFormat.format(toDate);
                return date + " | " + fromTime + " - " + toTime;
            }
        } catch (ParseException e) {
            e.printStackTrace();
        }
        return "";
    }

    private String getOrdinal(int number) {
        String[] suffixes = new String[]{"th", "st", "nd", "rd", "th", "th", "th", "th", "th", "th"};
        switch (number % 100) {
            case 11:
            case 12:
            case 13:
                return number + "th";
            default:
                return number + suffixes[number % 10];
        }
    }

    private String formatNoteDate(String dateTime) {
        try {
            SimpleDateFormat inputFormat =
                    new SimpleDateFormat("yyyy-MM-dd'T'HH:mm:ss", Locale.ENGLISH);

            SimpleDateFormat outputFormat =
                    new SimpleDateFormat("MMM dd, yyyy hh:mm a", Locale.ENGLISH);

            Date date = inputFormat.parse(dateTime);
            return outputFormat.format(date);

        } catch (Exception e) {
            e.printStackTrace();
        }
        return "";
    }

    //    private void setStatusBadgeStyle(TextView badge, String status) {
//        if (status == null) {
//            status = "Unknown";
//        }
//
//        String normalizedStatus = status.toLowerCase().trim();
//
//        if ("completed".equals(normalizedStatus)) {
//            badge.setText("Completed");
//            badge.setBackgroundResource(R.drawable.completed_badge);
//            badge.setTextColor(context.getResources().getColor(R.color.completed_text));
//
//        } else if ("cancelled".equals(normalizedStatus) || "canceled".equals(normalizedStatus)) {
//            badge.setText("Cancelled");
//            badge.setBackgroundResource(R.drawable.cancelled_badge);
//            badge.setTextColor(context.getResources().getColor(R.color.cancelled_text));
//
//        } else if ("upcoming".equals(normalizedStatus)|| "ongoing".equals(normalizedStatus)) {
//            badge.setText("Upcoming");
//            badge.setBackgroundResource(R.drawable.scheduled_badge);
//            badge.setTextColor(context.getResources().getColor(R.color.scheduled_text));
//
//        } else if ("payment_pending".equals(normalizedStatus)) {  // ⚠️ WITH UNDERSCORE
//            badge.setText("Payment Pending");
//            badge.setBackgroundResource(R.drawable.pending_badge);
//            badge.setTextColor(context.getResources().getColor(R.color.payment_pending_color));
//
//        } else {
//            // Default/unknown status
//            badge.setText(status);
//            badge.setBackgroundResource(R.drawable.scheduled_badge);
//            badge.setTextColor(context.getResources().getColor(R.color.scheduled_text));
//        }
//    }
    private void setStatusBadgeStyle(TextView badge, View timelineDot, String status) {
        if (status == null) {
            status = "Unknown";
        }

        String normalizedStatus = status.toLowerCase().trim();

        if ("completed".equals(normalizedStatus)) {
            // Badge styling
            badge.setText("Completed");
            badge.setBackgroundResource(R.drawable.completed_badge);
            badge.setTextColor(context.getResources().getColor(R.color.completed_text));

            // Timeline dot - GREEN
            timelineDot.setBackgroundResource(R.drawable.green_circular);

        } else if ("cancelled".equals(normalizedStatus) || "canceled".equals(normalizedStatus)) {
            // Badge styling
            badge.setText("Cancelled");
            badge.setBackgroundResource(R.drawable.cancelled_badge);
            badge.setTextColor(context.getResources().getColor(R.color.cancelled_text));

            // Timeline dot - RED
            timelineDot.setBackgroundResource(R.drawable.red_circular);

        } else if ("upcoming".equals(normalizedStatus) || "ongoing".equals(normalizedStatus)) {
            // Badge styling
            badge.setText("Upcoming");
            badge.setBackgroundResource(R.drawable.scheduled_badge);
            badge.setTextColor(context.getResources().getColor(R.color.scheduled_text));

            // Timeline dot - BLUE
            timelineDot.setBackgroundResource(R.drawable.circle_blue_bg);

        } else if ("payment_pending".equals(normalizedStatus)) {
            // Badge styling
            badge.setText("Payment Pending");
            badge.setBackgroundResource(R.drawable.pending_badge);
            badge.setTextColor(context.getResources().getColor(R.color.payment_pending_color));

            // Timeline dot - YELLOW
            timelineDot.setBackgroundResource(R.drawable.orange_circular);

        } else {
            // Default/unknown status
            badge.setText(status);
            badge.setBackgroundResource(R.drawable.scheduled_badge);
            badge.setTextColor(context.getResources().getColor(R.color.scheduled_text));

            // Timeline dot - GRAY (default)
            timelineDot.setBackgroundResource(R.drawable.circle_blue_bg);
        }
    }

    public static class ViewHolder extends RecyclerView.ViewHolder {
        View timelineDot, timelineLine;
        TextView tvAppointmentTitle, tvAppointmentDateTime, tvPaymentInfo, tvStatusBadge, tvCharCounter;
        AppCompatButton btnAddNote, btnCancelNote, btnSaveNote;
        LinearLayout llNoteSection;
        EditText etNoteInput;
        LinearLayout llNotesContainer;

        public ViewHolder(@NonNull View itemView) {
            super(itemView);
            llNotesContainer = itemView.findViewById(R.id.llNotesContainer);
            timelineDot = itemView.findViewById(R.id.timeline_dot);
            timelineLine = itemView.findViewById(R.id.timeline_line);
            tvAppointmentTitle = itemView.findViewById(R.id.tv_appointment_title);
            tvAppointmentDateTime = itemView.findViewById(R.id.tv_appointment_datetime);
            tvPaymentInfo = itemView.findViewById(R.id.tv_payment_info);
            tvStatusBadge = itemView.findViewById(R.id.tv_status_badge);
            tvCharCounter = itemView.findViewById(R.id.tv_char_counter);
            btnAddNote = itemView.findViewById(R.id.btn_add_note);
            btnCancelNote = itemView.findViewById(R.id.btn_cancel_note);
            btnSaveNote = itemView.findViewById(R.id.btn_save_note);
            llNoteSection = itemView.findViewById(R.id.ll_note_section);
            etNoteInput = itemView.findViewById(R.id.et_note_input);
        }
    }
}

File: app/src/main/java/com/digicoffer/lauditor/Appointments/Adapters/AppointmentHistoryAdapter.kt
package com.digicoffer.lauditor.Appointments.Adapters

import android.content.Context
import android.text.Editable
import android.text.TextWatcher
import android.view.LayoutInflater
import android.view.View
import android.view.ViewGroup
import android.widget.EditText
import android.widget.ImageView
import android.widget.LinearLayout
import android.widget.TextView
import androidx.annotation.NonNull
import androidx.appcompat.widget.AppCompatButton
import androidx.core.content.ContextCompat
import androidx.recyclerview.widget.RecyclerView
import com.digicoffer.lauditor.Appointments.Models.AppointmentModel
import com.digicoffer.lauditor.R
import com.digicoffer.lauditor.CommonFiles.GlobalFiles.AndroidUtils
import org.json.JSONObject
import java.text.ParseException
import java.text.SimpleDateFormat
import java.util.ArrayList
import java.util.Date
import java.util.Locale

class AppointmentHistoryAdapter(
    private val appointmentList: ArrayList<AppointmentModel>,
    private val context: Context,
    private val noteActionListener: OnNoteActionListener?
) : RecyclerView.Adapter<AppointmentHistoryAdapter.ViewHolder>() {

    var noteId: String = ""

    interface OnNoteActionListener {
        fun onSaveNote(appointment: AppointmentModel, note: String, position: Int)
        fun onCancelNote(appointment: AppointmentModel, position: Int)
        fun onEditNote(appointment: AppointmentModel, noteId: String, noteText: String, position: Int)
        fun onDeleteNote(appointment: AppointmentModel, noteId: String, position: Int)
    }

    override fun onCreateViewHolder(parent: ViewGroup, viewType: Int): ViewHolder {
        val view = LayoutInflater.from(parent.context)
            .inflate(R.layout.item_appointment_history, parent, false)
        return ViewHolder(view)
    }

    override fun onBindViewHolder(holder: ViewHolder, position: Int) {
        val appointment = appointmentList[position]

        // Hide timeline line for last item
        if (position == appointmentList.size - 1) {
            holder.timelineLine.visibility = View.INVISIBLE
        } else {
            holder.timelineLine.visibility = View.VISIBLE
        }

        // Set appointment title (count from end)
        val appointmentNumber = appointmentList.size - position
        val title = getOrdinal(appointmentNumber) + " Appointment"
        holder.tvAppointmentTitle.text = title

        // Format and set date/time
        val formattedDateTime = formatDateTime(appointment.appointment_from, appointment.appointment_to)
        holder.tvAppointmentDateTime.text = formattedDateTime

        // Set payment info
        if (appointment.payment != null) {
            val paymentInfo = "Payment: " + appointment.payment!!.label
            holder.tvPaymentInfo.text = paymentInfo
        }

        // Set status
        holder.tvStatusBadge.text = appointment.appointment_status
        // Set status badge AND timeline dot color
        setStatusBadgeStyle(holder.tvStatusBadge, holder.timelineDot, appointment.appointment_status)

        // Handle note section visibility
        if (appointment.isNoteExpanded) {
            holder.llNoteSection.visibility = View.VISIBLE
            holder.btnAddNote.visibility = View.GONE
            holder.etNoteInput.setText(appointment.tempNote)
            holder.etNoteInput.setSelection(holder.etNoteInput.text.length)
        } else {
            holder.llNoteSection.visibility = View.GONE
            holder.btnAddNote.visibility = View.VISIBLE
        }

        // Add Note button click
        holder.btnAddNote.setOnClickListener {
            val curPos = holder.bindingAdapterPosition
            if (curPos != RecyclerView.NO_POSITION && curPos >= 0 && curPos < appointmentList.size) {
                appointmentList[curPos].isNoteExpanded = true
                notifyItemChanged(curPos)
            }
        }

        holder.llNotesContainer.removeAllViews()

        if (appointment.isNoteExpanded) {
            // Editing Mode → Hide previous notes
            holder.llNotesContainer.visibility = View.GONE
            holder.btnAddNote.visibility = View.GONE
        } else if (appointment.notes != null && appointment.notes.length() > 0) {
            // Show previous notes only if NOT in edit mode
            holder.llNotesContainer.visibility = View.VISIBLE
            holder.btnAddNote.visibility = View.GONE

            for (i in 0 until appointment.notes.length()) {
                try {
                    val noteObj = appointment.notes.getJSONObject(i)
                    val noteText = noteObj.optString("note")
                    val createdOn = noteObj.optString("created_on")

                    val noteView = LayoutInflater.from(context)
                        .inflate(R.layout.appointments_notes_item, holder.llNotesContainer, false)

                    val tvNoteText: TextView = noteView.findViewById(R.id.tv_note_text)
                    val tvNoteDate: TextView = noteView.findViewById(R.id.tv_note_date)

                    tvNoteText.text = noteText
                    tvNoteDate.text = formatNoteDate(createdOn)
                    val ivEdit: ImageView = noteView.findViewById(R.id.iv_edit_note)
                    val ivDelete: ImageView = noteView.findViewById(R.id.iv_delete_note)

                    noteId = noteObj.optString("id")
                    val localNoteId = noteObj.optString("id")

                    ivEdit.setOnClickListener {
                        val curPos = holder.bindingAdapterPosition
                        if (curPos != RecyclerView.NO_POSITION && curPos >= 0 && curPos < appointmentList.size) {
                            val apt = appointmentList[curPos]
                            apt.isNoteExpanded = true
                            apt.isEditingNote = true
                            apt.editingNoteId = localNoteId
                            apt.tempNote = noteText
                            notifyItemChanged(curPos)
                        }
                    }

                    ivDelete.setOnClickListener {
                        val curPos = holder.bindingAdapterPosition
                        if (curPos != RecyclerView.NO_POSITION && curPos >= 0 && curPos < appointmentList.size) {
                            AndroidUtils.showConfirmationDialog(
                                context, "Delete Note", "Are you sure you want to delete this note?",
                                object : AndroidUtils.OnConfirmListener {
                                    override fun onSave() {
                                        noteActionListener?.onDeleteNote(appointmentList[curPos], localNoteId, curPos)
                                    }
                                    override fun onCancel() {}
                                }
                            )
                        }
                    }

                    holder.llNotesContainer.addView(noteView)
                } catch (e: Exception) {
                    e.printStackTrace()
                }
            }
        }

        AndroidUtils.ToggleButton(0, holder.btnSaveNote)

        // Character counter
        holder.etNoteInput.addTextChangedListener(object : TextWatcher {
            override fun beforeTextChanged(s: CharSequence?, start: Int, count: Int, after: Int) {}
            override fun onTextChanged(s: CharSequence?, start: Int, before: Int, count: Int) {
                val length = s?.length ?: 0
                holder.tvCharCounter.text = "$length/500"
                val curPos = holder.bindingAdapterPosition
                if (curPos != RecyclerView.NO_POSITION && curPos >= 0 && curPos < appointmentList.size) {
                    appointmentList[curPos].tempNote = s?.toString() ?: ""
                }
                if (s.isNullOrEmpty()) {
                    AndroidUtils.ToggleButton(0, holder.btnSaveNote)
                } else {
                    AndroidUtils.ToggleButton(1, holder.btnSaveNote)
                }
            }
            override fun afterTextChanged(s: Editable?) {}
        })

        // Cancel button
        holder.btnCancelNote.setOnClickListener {
            val curPos = holder.bindingAdapterPosition
            if (curPos != RecyclerView.NO_POSITION && curPos >= 0 && curPos < appointmentList.size) {
                val apt = appointmentList[curPos]
                apt.isNoteExpanded = false
                apt.tempNote = ""
                holder.etNoteInput.setText("")
                notifyItemChanged(curPos)

                noteActionListener?.onCancelNote(apt, curPos)
            }
        }

        holder.btnSaveNote.setOnClickListener {
            val curPos = holder.bindingAdapterPosition
            if (curPos != RecyclerView.NO_POSITION && curPos >= 0 && curPos < appointmentList.size) {
                val noteText = holder.etNoteInput.text.toString().trim()
                if (noteText.isEmpty()) return@setOnClickListener

                val apt = appointmentList[curPos]
                if (apt.isEditingNote) {
                    noteActionListener?.onEditNote(apt, apt.editingNoteId, noteText, curPos)
                } else {
                    noteActionListener?.onSaveNote(apt, noteText, curPos)
                }

                // Reset UI
                apt.isNoteExpanded = false
                apt.isEditingNote = false
                apt.editingNoteId = ""
                apt.tempNote = ""

                holder.etNoteInput.setText("")
                notifyItemChanged(curPos)
            }
        }
    }

    override fun getItemCount(): Int {
        return appointmentList.size
    }

    private fun formatDateTime(fromDateTime: String?, toDateTime: String?): String {
        if (fromDateTime.isNullOrEmpty() || toDateTime.isNullOrEmpty()) return ""
        try {
            val inputFormat = SimpleDateFormat("yyyy-MM-dd'T'HH:mm:ss", Locale.ENGLISH)
            val dateFormat = SimpleDateFormat("MMM dd, yyyy", Locale.ENGLISH)
            val timeFormat = SimpleDateFormat("hh:mm a", Locale.ENGLISH)

            val fromDate = inputFormat.parse(fromDateTime)
            val toDate = inputFormat.parse(toDateTime)

            if (fromDate != null && toDate != null) {
                val date = dateFormat.format(fromDate)
                val fromTime = timeFormat.format(fromDate)
                val toTime = timeFormat.format(toDate)
                return "$date | $fromTime - $toTime"
            }
        } catch (e: ParseException) {
            e.printStackTrace()
        }
        return ""
    }

    private fun getOrdinal(number: Int): String {
        val suffixes = arrayOf("th", "st", "nd", "rd", "th", "th", "th", "th", "th", "th")
        return when (number % 100) {
            11, 12, 13 -> "${number}th"
            else -> number.toString() + suffixes[number % 10]
        }
    }

    private fun formatNoteDate(dateTime: String?): String {
        if (dateTime.isNullOrEmpty()) return ""
        try {
            val inputFormat = SimpleDateFormat("yyyy-MM-dd'T'HH:mm:ss", Locale.ENGLISH)
            val outputFormat = SimpleDateFormat("MMM dd, yyyy hh:mm a", Locale.ENGLISH)
            val date = inputFormat.parse(dateTime)
            if (date != null) {
                return outputFormat.format(date)
            }
        } catch (e: Exception) {
            e.printStackTrace()
        }
        return ""
    }

    private fun setStatusBadgeStyle(badge: TextView, timelineDot: View, status: String?) {
        val normalizedStatus = status?.lowercase()?.trim() ?: "unknown"

        if ("completed" == normalizedStatus) {
            badge.text = "Completed"
            badge.setBackgroundResource(R.drawable.completed_badge)
            badge.setTextColor(ContextCompat.getColor(context, R.color.completed_text))
            timelineDot.setBackgroundResource(R.drawable.green_circular)
        } else if ("cancelled" == normalizedStatus || "canceled" == normalizedStatus) {
            badge.text = "Cancelled"
            badge.setBackgroundResource(R.drawable.cancelled_badge)
            badge.setTextColor(ContextCompat.getColor(context, R.color.cancelled_text))
            timelineDot.setBackgroundResource(R.drawable.red_circular)
        } else if ("upcoming" == normalizedStatus || "ongoing" == normalizedStatus) {
            badge.text = "Upcoming"
            badge.setBackgroundResource(R.drawable.scheduled_badge)
            badge.setTextColor(ContextCompat.getColor(context, R.color.scheduled_text))
            timelineDot.setBackgroundResource(R.drawable.circle_blue_bg)
        } else if ("payment_pending" == normalizedStatus) {
            badge.text = "Payment Pending"
            badge.setBackgroundResource(R.drawable.pending_badge)
            badge.setTextColor(ContextCompat.getColor(context, R.color.payment_pending_color))
            timelineDot.setBackgroundResource(R.drawable.orange_circular)
        } else {
            badge.text = status ?: "Unknown"
            badge.setBackgroundResource(R.drawable.scheduled_badge)
            badge.setTextColor(ContextCompat.getColor(context, R.color.scheduled_text))
            timelineDot.setBackgroundResource(R.drawable.circle_blue_bg)
        }
    }

    class ViewHolder(itemView: View) : RecyclerView.ViewHolder(itemView) {
        val timelineDot: View = itemView.findViewById(R.id.timeline_dot)
        val timelineLine: View = itemView.findViewById(R.id.timeline_line)
        val tvAppointmentTitle: TextView = itemView.findViewById(R.id.tv_appointment_title)
        val tvAppointmentDateTime: TextView = itemView.findViewById(R.id.tv_appointment_datetime)
        val tvPaymentInfo: TextView = itemView.findViewById(R.id.tv_payment_info)
        val tvStatusBadge: TextView = itemView.findViewById(R.id.tv_status_badge)
        val tvCharCounter: TextView = itemView.findViewById(R.id.tv_char_counter)
        val btnAddNote: AppCompatButton = itemView.findViewById(R.id.btn_add_note)
        val btnCancelNote: AppCompatButton = itemView.findViewById(R.id.btn_cancel_note)
        val btnSaveNote: AppCompatButton = itemView.findViewById(R.id.btn_save_note)
        val llNoteSection: LinearLayout = itemView.findViewById(R.id.ll_note_section)
        val etNoteInput: EditText = itemView.findViewById(R.id.et_note_input)
        val llNotesContainer: LinearLayout = itemView.findViewById(R.id.llNotesContainer)
    }
}
